compiler/rustc_middle/src/mir/query.rs

```diff
@@ -38,10 +38,81 @@ pub enum UnsafetyViolationDetails {
     AccessToUnionField,
     MutationOfLayoutConstrainedField,
     BorrowOfLayoutConstrainedField,
-    CallToFunctionWith,
+    CallToFunctionWith { missing_features: Vec<Symbol>, target_features: Vec<Symbol> },
 }
 
 impl UnsafetyViolationDetails {
+    fn missing_features(&self) -> Vec<Symbol> {
+        match self {
+            UnsafetyViolationDetails::CallToFunctionWith {
+                missing_features,
+                target_features: _,
+            } => missing_features.to_vec(),
+            _ => Vec::<Symbol>::new(),
+        }
+    }
+
+    fn target_features(&self) -> Vec<Symbol> {
```

Comment:
This should return `Option<Vec<Symbol>>` here instead of an empty vector. Then in the `note_missing_features` function, you don't need to check `is_empty` any more but can directly use `self.target_features()?`.